packages/snjs/lib/services/api/api_service.ts

```diff
@@ -2,50 +2,78 @@ import { UuidString } from './../../types';
   RevisionListEntry,
   RevisionListResponse,
-  SessionListResponse,
   SessionRenewalResponse,
   SignInResponse,
   SignOutResponse,
   SingleRevisionResponse,
   StatusCode,
   isErrorResponseExpiredToken,
+  ResponseMeta,
 } from './responses';
-import { Session, TokenSession } from './session';
+import { RemoteSession, Session, TokenSession } from './session';
 import { ContentType } from '@Models/content_types';
 import { PurePayload } from '@Payloads/pure_payload';
 import { SNRootKeyParams } from './../../protocol/key_params';
 import { SNStorageService } from './../storage_service';
-import { ErrorTag, HttpRequest, HttpVerb, SNHttpService } from './http_service';
+import {
+  ErrorTag,
+  HttpParams,
+  HttpRequest,
+  HttpVerb,
+  SNHttpService,
+} from './http_service';
 import merge from 'lodash/merge';
 import { ApiEndpointParam } from '@Services/api/keys';
 import * as messages from '@Services/api/messages';
 import { PureService } from '@Services/pure_service';
-import { joinPaths } from '@Lib/utils';
+import { isNullOrUndefined, joinPaths } from '@Lib/utils';
 import { StorageKey } from '@Lib/storage_keys';
-
-const REQUEST_PATH_KEY_PARAMS = '/auth/params';
-const REQUEST_PATH_REGISTER = '/auth';
-const REQUEST_PATH_LOGIN = '/auth/sign_in';
-const REQUEST_PATH_CHANGE_PW = '/auth/change_pw';
-const REQUEST_PATH_SYNC = '/items/sync';
-const REQUEST_PATH_LOGOUT = '/auth/sign_out';
-const REQUEST_PATH_SESSION_REFRESH = '/session/refresh';
-const REQUEST_PATH_ALL_SESSIONS = '/sessions';
-const REQUEST_PATH_SESSION = '/session';
-const REQUEST_PATH_ITEM_REVISIONS = '/items/:item_id/revisions';
-const REQUEST_PATH_ITEM_REVISION = '/items/:item_id/revisions/:id';
-
-const API_VERSION = '20200115';
+import { SNPermissionsService } from '../permissions_service';
+
+type PathNames = {
+  keyParams: string;
+  register: string;
+  signIn: string;
+  changePassword: string;
+  sync: string;
+  signOut: string;
+  refreshSession: string;
+  sessions: string;
+  session: string;
+  itemRevisions: (itemId: string) => string;
+  itemRevision: (itemId: string, revisionId: string) => string;
+};
+
+const Paths: {
+  v0: PathNames
+} = {
+  v0: {
+    keyParams: '/auth/params',
+    register: '/auth',
+    signIn: '/auth/sign_in',
+    changePassword: '/auth/change_pw',
+    sync: '/items/sync',
+    signOut: '/auth/sign_out',
+    refreshSession: '/session/refresh',
+    sessions: '/sessions',
+    session: '/session',
+    itemRevisions: (itemId: string) => `/items/${itemId}/revisions`,
+    itemRevision: (itemId: string, revisionId: string) =>
+      `/items/${itemId}/revisions/${revisionId}`,
+  },
+};
 
 type InvalidSessionObserver = (revoked: boolean) => void;
 
 export class SNApiService extends PureService {
-  private host?: string;
+  /** Legacy api version field to be specified in params when calling v0 APIs. */
+  static readonly V0_API_VERSION = '20200115';
+
   private session?: Session;
+  private host!: string;
```

Comment:
is the bang operator necessary here?

Reply:
Hm, I should probably just make `this.host` a required parameter in the constructor yeah…